### PR TITLE
Disable duplicate routes in the router

### PR DESCRIPTION
**Pull request by smarterclayton**
Only the oldest route for a given hostname is used.

Explore some alternative options for implementation.

@pweil- @ramr

Added one commit on top of #4421

### CODE REVIEW

File: images/origin/Dockerfile [hunk 1]
```diff
@@ -17,7 +17,7 @@ RUN ln -s /usr/bin/openshift /usr/bin/oc && \
     ln -s /usr/bin/openshift /usr/bin/openshift-deploy && \
     ln -s /usr/bin/openshift /usr/bin/openshift-docker-build && \
     ln -s /usr/bin/openshift /usr/bin/openshift-sti-build && \
-    ln -s /usr/bin/openshift /usr/bin/openshift-router && \
+    ln -s /usr/bin/openshift /usr/bin/openshift-f5-router && \
```
Comment (ramr) [inline images/origin/Dockerfile]: Was this intentional - to change ```openshift-router``` to ```openshift-f5-router```?
I guess the main question really is why was this needed the origin dockerfile? Both the f5 and haproxy routers run inside a container and have the appropriate router binary in their images.

Comment (ramr) [inline images/origin/Dockerfile]: Ok - see why you would need it if you removed it from the f5 dockerfile. imo, it feels better to remove it here and just keep it in the f5 dockerfile - its consistent with the haproxy one.

File: images/router/f5/Dockerfile [hunk 2]
```diff
@@ -1,11 +1,8 @@
...
 # The standard name for this image is openshift/origin-f5-router
 #
-FROM openshift/origin-base
+FROM openshift/origin
 
-ADD bin/openshift /usr/bin/openshift
-RUN ln -s /usr/bin/openshift /usr/bin/openshift-router
-
-ENTRYPOINT ["/usr/bin/openshift-router", "--type=f5"]
+ENTRYPOINT ["/usr/bin/openshift-f5-router"]
```
Comment (ramr) [inline images/router/f5/Dockerfile]: :+1:  for a cleaner dockerfile - but see above comment on moving the symlink for ```openshift-f5-router```.

Comment (smarterclayton) [inline images/router/f5/Dockerfile]: Hrm - all of our symlinks for openshift/origin are in the base image - I'd
rather have it there.  The router image is distinct enough that I don't
think it's considered the same.  If we switched the f5 to a separate binary
we'd want to, but for now I'd rather keep them all together.

On Mon, Aug 31, 2015 at 3:44 PM, Ram Ranganathan <notifications@github.com>
wrote:

> In images/router/f5/Dockerfile
> <https://github.com/openshift/origin/pull/4428#discussion_r38344729>:
>
> >
> > -ADD bin/openshift /usr/bin/openshift
> > -RUN ln -s /usr/bin/openshift /usr/bin/openshift-router
> > -
> > -ENTRYPOINT ["/usr/bin/openshift-router", "--type=f5"]
> > +ENTRYPOINT ["/usr/bin/openshift-f5-router"]
>
> [image: :+1:] for a cleaner dockerfile - but see above comment on moving
> the symlink for openshift-f5-router.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4428/files#r38344729>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

Comment (ramr) [inline images/router/f5/Dockerfile]: No strong opinion either way but it just felt odd to me - to have all those symlinks stuffed in origin-base ... imho, would have gone w/ a minimalistic approach - keep the base set small and let the specializations take do/whatever they need to [it helps with the layered build approach as well].

Comment (smarterclayton) [inline images/router/f5/Dockerfile]: It makes all of the other images one layer smaller which has a material
cost during pull.

On Aug 31, 2015, at 5:50 PM, Ram Ranganathan <notifications@github.com>
wrote:

In images/router/f5/Dockerfile
<https://github.com/openshift/origin/pull/4428#discussion_r38364013>:

>
> -ADD bin/openshift /usr/bin/openshift
> -RUN ln -s /usr/bin/openshift /usr/bin/openshift-router
> -
> -ENTRYPOINT ["/usr/bin/openshift-router", "--type=f5"]
> +ENTRYPOINT ["/usr/bin/openshift-f5-router"]

No strong opinion either way but it just felt odd to me - to have all those
symlinks stuffed in origin-base ... imho, would have gone w/ a minimalistic
approach - keep the base set small and let the specializations take
do/whatever they need to [it helps with the layered build approach as well].

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4428/files#r38364013>.

File: pkg/cmd/infra/router/template.go [hunk 3]
```diff
@@ -2,87 +2,162 @@ package router
...
+
+	return cmd
+}
+
+func (o *TemplateRouterOptions) Complete() error {
+	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
+	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
+	if len(routerSvcName) > 0 {
+		if len(routerSvcNamespace) == 0 {
+			return fmt.Errorf("ROUTER_SERVICE_NAMESPACE is required when ROUTER_SERVICE_NAME is specified")
```
Comment (ramr) [inline pkg/cmd/infra/router/template.go]: Any reason this code is in ```Complete``` vis-a-vis in ```Validate``` below (as it is a validation)?

Comment (smarterclayton) [inline pkg/cmd/infra/router/template.go]: Things we pull from Env come from Complete

On Mon, Aug 31, 2015 at 7:07 PM, Ram Ranganathan <notifications@github.com>
wrote:

> In pkg/cmd/infra/router/template.go
> <https://github.com/openshift/origin/pull/4428#discussion_r38365620>:
>
> > +	cmd.AddCommand(version.NewVersionCommand(name))
> > +
> > +	flag := cmd.Flags()
> > +	options.Config.Bind(flag)
> > +	options.TemplateRouter.Bind(flag)
> > +	options.RouterSelection.Bind(flag)
> > +
> > +	return cmd
> > +}
> > +
> > +func (o *TemplateRouterOptions) Complete() error {
> > +	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
> > +	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
> > +	if len(routerSvcName) > 0 {
> > +		if len(routerSvcNamespace) == 0 {
> > +			return fmt.Errorf("ROUTER_SERVICE_NAMESPACE is required when ROUTER_SERVICE_NAME is specified")
>
> Any reason this code is in Complete vis-a-vis in Validate below (as it is
> a validation)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4428/files#r38365620>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

File: pkg/router/controller/factory/factory.go [hunk 4]
```diff
@@ -49,26 +82,133 @@ func (factory *RouterControllerFactory) Create(plugin router.Plugin) *controller
...
+			}
+			changed()
+		}
+	}, time.Second, util.NeverStop)
+	go util.Until(func() {
+		for {
+			if _, _, err := endpointsEventQueue.Pop(); err != nil {
+				return
+			}
+			changed()
```
Comment (ramr) [inline pkg/router/controller/factory/factory.go]: Would be good to add a parameter to the callback ```changed``` to indicate what changed - endpoint/route. 
This doesn't seem to be used currently, right?

Comment (smarterclayton) [inline pkg/router/controller/factory/factory.go]: Yeah I'm playing around with this for a future index based store (where the
template router gets a lot simpler).  I want to queue up changes to the
config over several seconds before applying them, and also get better at
identifying when things actually change.  However it's not ready yet.

On Aug 31, 2015, at 7:49 PM, Ram Ranganathan <notifications@github.com>
wrote:

In pkg/router/controller/factory/factory.go
<https://github.com/openshift/origin/pull/4428#discussion_r38373277>:

> +	}, &kapi.Endpoints{}, endpointsEventQueue, factory.ResyncInterval).Run()
> +
> +	go util.Until(func() {
> +		for {
> +			if _, _, err := routeEventQueue.Pop(); err != nil {
> +				return
> +			}
> +			changed()
> +		}
> +	}, time.Second, util.NeverStop)
> +	go util.Until(func() {
> +		for {
> +			if _, _, err := endpointsEventQueue.Pop(); err != nil {
> +				return
> +			}
> +			changed()

Would be good to add a parameter to the callback changed to indicate what
changed - endpoint/route.
This doesn't seem to be used currently, right?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4428/files#r38373277>.

**Comment by smarterclayton** [comment]
[test], comments addressed (but the other pull needs to be reviewed and merged first)

**Comment by smarterclayton** [comment]
Other review comments?  I need to sort out the test failure and add an explicit integration test... But what else?

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Oh, I didn't properly handle paths with overlapping hosts.  Good test cases Ram and Paul :)

So here's the solution:

1. The oldest route on a particular host wins - if another route exists that is newer with the same host, it is ignored
2. Only subpaths from the same namespace as the oldest route can be used.
3. If the oldest route on a particular host is deleted, the next oldest route becomes the lead and conflicting routes are no longer available.

Example:

* Route alpha/first "www.example.com" T0
* Route beta/first "www.example.com" T1
* Route alpha/second "www.example.com/path" T2
* Route beta/second "www.example.com/path" T3
* Route beta/third "www.example.com/path2" T4

Only alpha/first and alpha/second are used (alpha/first is oldest, alpha/second is in the same namespace).  If we delete alpha/first, then beta/first, beta/second, and beta/third are used (beta/first is oldest).

However, because it's more complex to switch to the "hold everything, only expose what is used model" with the current design of the template plugin and router, I'm only going to implement filtering, so that *all* previous claims are removed if a new claim comes in that is older than all previous claims.  The next resync cycle will recalculate the correct route.  This puts more emphasis on moving to the second model I proposed, but it's still a bigger change and I don't want to get bogged down in retesting everything there.

**Comment by smarterclayton** [comment]
@ramr can you sign off on this?

**Comment by ramr** [comment]
LGTM

**Comment by smarterclayton** [comment]
[merge]

**Comment by sosiouxme** [comment]
Just for the record, I asked about this on IRC:

pweil: sosiouxme, there were some enhancements so that the router only serves unique routes.  Oldest version wins. I think Ram was going to put in some docs on that
sosiouxme:  pweil: thanks for the pointer to the PR. so i see two shortcomings with that. (1) only applies to our haproxy router and (2) if you happen to recreate your route (say to update the certificate) then someone else who's created a route to camp your domain (or perhaps by accident) suddenly is the oldest route and steals your traffic. do you know of any plans for preventing duplicates in the first place?
claytonc: sosiouxme: we will not
claytonc: because in the real world you cannot
claytonc: you shouldn't have to recreate your route to update the cert
claytonc: we can register that you were rejected and record that for the future
claytonc: but that's not going to work with multiple subdomained routers in the future
sosiouxme: you don't have to recreate your route, but you might if you're not aware of the potential consequences
claytonc: sosiouxme: so the counter argument
claytonc: i'm user 1
claytonc: i want to use www.foo.com in project bar
claytonc: i delete it, then recreate it in another project
claytonc: do i have to contact an admin to use www.foo.com now?
sosiouxme: ok that's a better use case
claytonc: the answer may be yes - that would be having the router say "i've seen this host before, therefore you get some boolean on your route that has to be cleared by an admin"
claytonc: we can do that
claytonc: but it also means you can't live migrate hosts across projcets
